### PR TITLE
Define semantics of avio brand for image items

### LINKED ISSUE #105: Clarify the use of 'avio' brand for multi-layer images
If 'avio' is set, no image item shall use layered images.

### PR DESCRIPTION
**Pull request by cconcolato**
closes #105

***
<a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/av1-avif/pull/114.html" title="Last updated on Jan 6, 2021, 1:48 AM UTC (0ccebed)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/av1-avif/114/3fb01d0...0ccebed.html" title="Last updated on Jan 6, 2021, 1:48 AM UTC (0ccebed)">Diff</a>

### CODE REVIEW

**Comment by leo-barnes** [comment]
Is `avio` defined somewhere else except in the AVIF spec? I only see it mentioned with `should`s in AVIF. Don't we need to have a `shall` somewhere to say that it may not be used if the primary item/sequence contains predicted frames? Or maybe that's unnecessarily strict.

**Comment by cconcolato** [comment]
`avio` is only defined in AVIF. I agree to rephrase it with shalls.

**Review by leo-barnes — APPROVED** [review]
Looks good to me!

File: index.bs [hunk 1]
```diff
@@ -147,7 +152,8 @@ This specification reuses syntax and semantics used in [[!AV1-ISOBMFF]].
 
 <h3 id="image-item">AV1 Image Item</h3>
 
-When an image item is of type <dfn value export for="AV1 Image Item Type">av01</dfn>, it is called an <dfn>AV1 Image Item</dfn>, and shall obey the following constraints:
+When an item is of type <dfn value export for="AV1 Image Item Type">av01</dfn>, it is called an <dfn>AV1 Image Item</dfn>, and shall obey the following constraints:
```
Comment (wantehchang) [inline index.bs:155]: Just wanted to confirm that you meant to delete the word "image" from "When an image item is of type ..."

File: index.bs [hunk 2]
```diff
@@ -189,66 +195,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
 <p>
     An <dfn>AV1 Image Sequence</dfn> is defined as a set of AV1 [=Temporal Units=] stored in an [=AV1 track=] as defined in [[!AV1-ISOBMFF]] with the following constraints:
 
+    - The track shall be a valid [=MIAF image sequence=].
     - The track handler for an [=AV1 Image Sequence=] shall be <code>pict</code>.
 
     <h2 id="auxiliary-images">Auxiliary Image Items and Sequences</h2>
 <p>An <dfn>AV1 Auxiliary Image Item</dfn> (respectively an <dfn>AV1 Auxiliary Image Sequence</dfn>) is an [=AV1 Image Item=] (respectively [=AV1 Image Sequence=]</dfn>) with the following additional constraints:
+    - It shall be a compliant [=MIAF Auxiliary Image Item=] (respectively [=MIAF Auxiliary Image Sequence=])
```
Comment (wantehchang) [inline index.bs:203]: Nit: Add a period (.) at the end of this sentence.

File: index.bs [hunk 3]
```diff
@@ -189,66 +195,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
...
 
   <h2 id="brands">Brands, Internet media types and file extensions</h2>
 
 <h3 id="brands-overview">Brands overview</h3>
 
-<p>As defined by [[!ISOBMFF]], the presence of a brand in the <code>FileTypeBox</code> (respectively <code>TrackTypeBox</code>) can be interpreted as the permission for those [=AV1 Image File Format=] readers/parsers and [=AV1 Image File Format=] renderers that only implement the features required by the brand, to process the corresponding file (respectively item or sequence).</p>
+<p>As defined by [[!ISOBMFF]], the presence of a brand in the <code>[=compatible_brands=]</code> list in the <code>[=FileTypeBox=]</code> can be interpreted as the permission for those [=AV1 Image File Format=] readers/parsers and [=AV1 Image File Format=] renderers that only implement the features required by the brand, to process the corresponding file and only the parts (e.g. items or sequences) that comply with the brand.</p>
```
Comment (wantehchang) [inline index.bs:217]: Should this also describe the presence of a brand in the major_brand field?

File: index.bs [hunk 4]
```diff
@@ -189,66 +195,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
...
-- the brand <dfn value="" export="" for="AVIF Image brand">avif</dfn> ,
-- the brand 'mif1' as recommended in [[!HEIF]],
-- a brand to identify the AVIF profile (see section [[#profiles]]), if any, with which the file complies.
+The brand to identify [=AV1 image items=] is <dfn value="avif" export="" for="AVIF Image brand">avif</dfn>.
+
+Files that indicate this brand shall comply with the following:
+- there shall be an [=AV1 image item=] among the set of items comprised of the primary item and its alternates.
+- [=AV1 auxiliary image items=] as specified in [[#auxiliary-images]] may be present in the file.
+
+Additionally, the brand <dfn value="avio" export="" for="AVIF Intra-only brand">avio</dfn> is defined. If the file indicates the brand <code>[=avio=]</code> in the [=compatible_brands=] field of the [=FileTypeBox=], then there shall be an [=AV1 image item=] made only of [=Key Frames=] among the set of items comprised of the primary item and its alternates. Conversely, if such item exists, the brand <code>[=avio=]</code> should be used.
```
Comment (wantehchang) [inline index.bs:230]: Nit: such item => such an item

File: index.bs [hunk 5]
```diff
@@ -189,66 +195,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
...
-- a brand to identify the AVIF profile (see section [[#profiles]]), if any, with which the file complies.
+The brand to identify AVIF image sequences is <dfn value="avis" export="" for="AVIF Image Sequence brand">avis</dfn>.
+
+Files that indicate this brand shall comply with the following:
+- they shall contain one or more [=AV1 image sequences=] as defined in [[#image-sequences]].
+- they may contain [=AV1 auxiliary image sequences=] as defined in [[#auxiliary-images]].
 
-Additionally, if the image sequences are made only of [=AV1 Samples=] marked as 'sync', then the brand <dfn value="" export="" for="AVIF Intra-only brand">avio</dfn> should be used.
+Files that conform with these constraints should include in the [=compatible_brands=] field of the [=FileTypeBox=]:
+- the brand <code>'[=avis=]'</code>,
```
Comment (wantehchang) [inline index.bs:240]: Nit/optional: If the major_brand field is "avis", it is not clear whether this sentence implies that "avis" should appear in the compatible_brands field again.

File: index.bs [hunk 6]
```diff
@@ -189,66 +195,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
...
+
+Additionally, if a file contains [=AV1 image sequences=] and the brand <code>[=avio=]</code> is used in the [=compatible_brands=] field of the [=FileTypeBox=], the item constraints for this brand shall be met and at least one of the [=AV1 image sequences=] shall be made only of [=AV1 Samples=] marked as 'sync'. Conversely, if such track exists and the constraints of the brand <code>[=avio=]</code> on [=AV1 image items=] are met, the brand should be used.
```
Comment (wantehchang) [inline index.bs:242]: Nit: such track => such a track

File: index.bs [hunk 7]
```diff
@@ -189,66 +195,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
...
 
 NOTE: As defined in [[!MIAF]], a file that is primarily an image sequence still has at least a primary image item. Hence, it can also declare brands for signaling the image item.
 
+  <h2 id="file-constraints">File constraints</h3>
+
+  The following constraints are common to files compliant to this specification:
```
Comment (wantehchang) [inline index.bs:248]: Nit/optional: compliant to => compliant with

File: index.bs [hunk 8]
```diff
@@ -189,66 +195,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
...
+  The [=FileTypeBox=] should declare at least one profile that enables decoding of the primary image item, or one of its alternates. It is not an error for the encoder to include an auxiliary image that is not allowed by the specified profile(s).
 
-  <h3 id="profiles-constraints">Common constraints</h3>
-
-  The following constraints are common to all profiles defined in this specification:
-  - The file shall be compliant with the [[!MIAF]] specification and list 'miaf' in the [=compatible_brands=] field of the [=FileTypeBox=].
-  - The [=primary image=], or one of its alternates, shall be an [=AV1 Image Item=] or be a derived image that references one or more items that all are [=AV1 Image Items=].
+  If an image sequence brand is declared in the [=FileTypeBox=], if a profile is declared in the [=FileTypeBox=], the profile shall also enable decoding of at least one image sequence track. The profile should allow decoding of any associated auxiliary image sequence tracks, unless it is acceptable to decode the image sequence without its auxiliary image sequence tracks.
```
Comment (wantehchang) [inline index.bs:261]: Nit: In the first sentence, I think the second "if" could be replaced with "and".

File: index.bs [hunk 9]
```diff
@@ -264,19 +280,19 @@ A file containing items compliant with this profile is expected to list the foll
 
 A file containing a 'pict' track compliant with this profile is expected to list the following brands, in any order, in the [=compatible_brands=] of the [=FileTypeBox=]:
 
-  <code>avis, msf1, miaf, MA1B</code>
+  <code>avis, avif, msf1, miaf, MA1B</code>
```
Comment (wantehchang) [inline index.bs:283]: [Just a comment, not a request for change] Listing both "avis" and "avif" increases the image container overhead. Container overhead is a known problem with AVIF. It seems bad to make the problem worse.

**Review by wantehchang — APPROVED** [review]
LGTM. Thank you very much for writing these changes.

**Comment by leo-barnes** [comment]
Per meeting, the conclusion was that we wanted the following changes:
1. Infer `avif` from `avis` brand to keep down the header overhead.
2. Remove `or its alternates` with regards to the primary item when using the `avif` brand.
3. Remove usage of TrackTypeBox.

File: index.bs (hunk 1 above)
Comment (cconcolato) [reply index.bs:156]: Yes, 'av01' implies that it is an image item.

File: index.bs (hunk 3 above)
Comment (cconcolato) [reply index.bs:218]: ISOBMFF guides how to write derived specification and says:
>do not claim the major_brand field (“files conformant to this specification must set the major_brand to XXXX”) as a file could never be conformant to two such specifications written that way, and it also blocks someone also from deriving a specification from yours. However, brands that are only permitted as compatible brands may be defined.

File: index.bs (hunk 5 above)
Comment (cconcolato) [reply index.bs]: It is a common practice to replicate the major brand in the compatible brands, but I will double check the impact of not doing it.

File: index.bs [hunk 10]
```diff
@@ -120,6 +124,8 @@ url: #biblio-av1; spec: AV1; type: dfn;
     text: Sequence Header OBU
     text: Metadata OBU
     text: Temporal Unit
+    text: Key Frame
+    text: Intra Frame
```
Comment (wantehchang) [inline index.bs:128]: These frame types are called KEY_FRAME and INTRA_ONLY_FRAME respectively in AV1. So "Intra Frame" probably should be changed to "Intra Only Frame".

File: index.bs [hunk 11]
```diff
@@ -189,66 +196,76 @@ For multi-layered images, the LayerSelectorProperty may be present as defined in
...
+The brand to identify [=AV1 image items=] is <dfn value="avif" export="" for="AVIF Image brand">avif</dfn>.
+
+Files that indicate this brand in the [=compatible_brands=] field of the [=FileTypeBox=] shall comply with the following:
+- The primary item shall be an [=AV1 Image Item=] or be a derived image that references one or more items that all are [=AV1 Image Items=].
+- [=AV1 auxiliary image items=] may be present in the file.
+
+Files that conform with these constraints should include the brand <code>[=avif=]</code> in the [=compatible_brands=] field of the [=FileTypeBox=].
+
+Additionally, the brand <dfn value="avio" export="" for="AVIF Intra-only brand">avio</dfn> is defined. If the file indicates the brand <code>[=avio=]</code> in the [=compatible_brands=] field of the [=FileTypeBox=], then the primary item or all the items referenced by the primary item shall be [=AV1 image items=] made only of [=Key Frames=] or [=Intra Frames=]. Conversely, if the previous constraint applies, the brand <code>[=avio=]</code> should be used in the [=compatible_brands=] field of the [=FileTypeBox=].
```
Comment (wantehchang) [inline index.bs:233]: As noted above, "Intra Frames" probably should be changed to "Intra Only Frames".

**Review by wantehchang — APPROVED** [review]

File: index.bs (hunk 10 above)
Comment (cconcolato) [reply index.bs:128]: The definition section in the AV1 spec says:
> Intra frame
A frame compressed using only intra prediction which can be independently decoded.

I think the all caps name refer to `frame_type` values.

**Review by leo-barnes — APPROVED** [review]
LGTM!

File: index.bs (hunk 10 above)
Comment (wantehchang) [reply index.bs:128]: I was just being pedantic about terminology. I think that an intra frame in AV1 is either a KEY_FRAME or an INTRA_ONLY_FRAME. This claim is justified by the following two descriptions in the AV1 spec:
- In Section 2 (Terms and definitions), a key frame is defined as "An Intra frame which resets the decoding process when it is shown."
- In Section 5.9.2 (Uncompressed header syntax), the `FrameIsIntra` variable is defined as `FrameIsIntra = (frame_type == INTRA_ONLY_FRAME || frame_type == KEY_FRAME)`.

So I propose either of the following changes:
1. Remove "Key Frame" and just use the term "Intra Frame" to refer to both KEY_FRAME and INTRA_ONLY_FRAME.
2. Change "Intra Frame" to "Intra Only Frame".

I prefer the first change.

**Comment by cconcolato** [comment]
@wantehchang I think I addressed your last comments. Let me know otherwise.

**Review by wantehchang — APPROVED** [review]
LGTM. Thank you, Cyril.

**Pull request merged**